src/components/table/AposTable.vue

```diff
@@ -0,0 +1,201 @@
+              v-bind:is="icons[header.icon]"
+            ></component>
+            {{ header.label }}
+          </component>
+        </th>
+      </tr>
+      <AposTableRow :headers="headers" 
+        v-for="row in rows" 
+        :row="row" v-bind:key="row.id"
+        :allChecked="allChecked"
+        :someChecked="someChecked"
+        v-on:checked="checkHandler"
+      />
+    </tbody>
+  </table>  
+</template>
+
+<script>
+import AposTableRow from './AposTableRow.vue';
+import AposCheckboxInput from '../inputCheckbox/AposCheckboxInput.vue';
+
+export default {
+  components: {
+    AposTableRow,
+    AposCheckboxInput
+  },
+  props: {
+    headers: {
+      type: Array,
+      required: true
+    },
+    rows: {
+      type: Array,
+      required: true
+    }
+  },
+  data() {
+    // fetch all icons used in table headers
+    const icons = {};
+    this.headers.forEach(h => {
+      if (h.icon) {
+        icons[h.icon] = () => import(`vue-material-design-icons/${h.icon}.vue`);
+      }
+    });
+    return {
+      icons,
+      allChecked: false,
+      someChecked: false,
+      checked: [],
+      selectAllField: {
+        status: {},
+        value: { data: [] },
+        field: {
+          name: 'selectAll',
+          type: 'checkbox',
+          choices: [ { value: 'checked' } ]
+        } 
+      }
+    }
+  },
+  mounted() {
+  },
+  methods: {
+
+    // handle state changes caused by checking and unchecking rows
+    checkHandler(id, checked) {
+      if (checked) {
+        this.checked.push(id);
+      }
+      if (!checked) {
+        this.checked = this.checked.filter(checkedId => checkedId !== id);
+      }
+
+      if (this.checked.length === 0) {
+        this.selectAllField.value.data = [];
+        this.someChecked = false;
+        this.allChecked = false;
+      }
+
+      if (this.checked.length === this.rows.length) {
+        this.allChecked = true;
+        if (this.selectAllField.field.choices[0].indeterminate) {
+          delete this.selectAllField.field.choices[0].indeterminate;
+        }
+        // don't know why i have to flush this out so hard
+        this.selectAllField.value.data = [];
+        this.selectAllField.value.data = ['checked'];
+      }
+
+      if (!checked && this.allChecked) {
+        this.someChecked = true;
+        this.allChecked = false;
+        this.selectAllField.field.choices[0].indeterminate = true;
+        // don't know why i have to flush this out so hard
+        this.selectAllField.value.data = [];
+        this.selectAllField.value.data = ['checked'];
+      }
+    },
+    selectAll(response) {
+      const checked = response.data.length > 0 ? true : false;
+      const indeterminate = this.selectAllField.field.choices[0].indeterminate;
+
+      // if selectAll is true but also indeterminate, bail out
+      // the checkbox has to be checked to light up the indeterminate ui
+      if (checked && indeterminate) {
+        return;
```

Comment:
I'm having trouble following the indeterminate thread, but shouldn't it go indeterminate once a single row is checked? That's also the GMail UX.

Reply:
yes, i'll add that